File: packages/server/cmd/renewlet/ai_models_test.go
package main

// AI 模型列表测试保护 Go 代理对 OpenAI/Gemini/Anthropic 三类响应的归一化、超时和原始错误回显契约。
// 这些断言要和 Cloudflare Worker 的模型列表测试保持语义一致，避免两端设置页候选行为漂移。
import (
	"context"
	"errors"
	"net/http"
	"net/http/httptest"
	"strings"
	"testing"
	"time"
)

func withAIModelListTestServer(t *testing.T, handler http.HandlerFunc) *httptest.Server {
	t.Helper()
	server := httptest.NewServer(handler)
	previousClient := aiModelListHTTPClient
	aiModelListHTTPClient = server.Client()
	t.Cleanup(func() {
		aiModelListHTTPClient = previousClient
		server.Close()
	})
	return server
}

func TestAIModelListOpenAIShape(t *testing.T) {
	server := withAIModelListTestServer(t, func(w http.ResponseWriter, r *http.Request) {
		if r.URL.Path != "/v1/models" {
			t.Fatalf("unexpected path: %s", r.URL.Path)
		}
		if got := r.Header.Get("Authorization"); got != "Bearer sk-test-secret" {
			t.Fatalf("unexpected authorization header: %q", got)
		}
		w.Header().Set("Content-Type", "application/json")
		_, _ = w.Write([]byte(`{"data":[{"id":"gpt-5.1","created":1780000000,"owned_by":"openai"},{"id":"gpt-5.1"}]}`))
	})

	response, err := listAIModels(context.Background(), aiModelListRequest{
		ProviderType: aiProviderTypeOpenAI,
		BaseURL:      server.URL + "/v1",
		APIKey:       "sk-test-secret",
	}, localeZhCN)
	if err != nil {
		t.Fatal(err)
	}
	if len(response.Models) != 1 || response.Models[0].ID != "gpt-5.1" {
		t.Fatalf("unexpected models: %#v", response.Models)
	}
	if response.ProviderType != aiProviderTypeOpenAI || response.TransportProtocol != aiProtocolOpenAIChat {
		t.Fatalf("provider/protocol not preserved: %#v", response)
	}
	if response.Models[0].OwnedBy == nil || *response.Models[0].OwnedBy != "openai" {
		t.Fatalf("owner not parsed: %#v", response.Models[0])
	}
}

func TestAIModelListGeminiShape(t *testing.T) {
	server := withAIModelListTestServer(t, func(w http.ResponseWriter, r *http.Request) {
		if got := r.Header.Get("x-goog-api-key"); got != "AIza-test-secret" {
			t.Fatalf("unexpected Gemini key header: %q", got)
		}
		_, _ = w.Write([]byte(`{"models":[{"name":"models/gemini-2.5-pro","baseModelId":"gemini-2.5-pro","displayName":"Gemini 2.5 Pro","inputTokenLimit":1048576,"outputTokenLimit":65536,"supportedGenerationMethods":["generateContent"],"thinking":true},{"name":"models/text-embedding","supportedGenerationMethods":["embedContent"]}]}`))
	})

	response, err := listAIModels(context.Background(), aiModelListRequest{
		ProviderType: aiProviderTypeGemini,
		BaseURL:      server.URL + "/v1beta",
		APIKey:       "AIza-test-secret",
	}, localeZhCN)
	if err != nil {
		t.Fatal(err)
	}
	if len(response.Models) != 1 || response.Models[0].ID != "gemini-2.5-pro" {
		t.Fatalf("unexpected Gemini models: %#v", response.Models)
	}
	if response.Models[0].Capabilities.Thinking == nil || !*response.Models[0].Capabilities.Thinking {
		t.Fatalf("thinking capability not parsed: %#v", response.Models[0].Capabilities)
	}
}

func TestAIModelListAnthropicShape(t *testing.T) {
	server := withAIModelListTestServer(t, func(w http.ResponseWriter, r *http.Request) {
		if r.URL.Path != "/v1/models" {
			t.Fatalf("unexpected path: %s", r.URL.Path)
		}
		if got := r.Header.Get("x-api-key"); got != "sk-ant-test-secret" {
			t.Fatalf("unexpected Anthropic key header: %q", got)
		}
		if got := r.Header.Get("anthropic-version"); got != "2023-06-01" {
			t.Fatalf("unexpected Anthropic version header: %q", got)
		}
		_, _ = w.Write([]byte(`{"data":[{"id":"claude-sonnet-4-6","display_name":"Claude Sonnet 4.6","created_at":"2026-01-01T00:00:00Z","type":"model"}]}`))
	})

	response, err := listAIModels(context.Background(), aiModelListRequest{
		ProviderType: aiProviderTypeAnthropic,
		BaseURL:      server.URL + "/v1",
		APIKey:       "sk-ant-test-secret",
	}, localeZhCN)
	if err != nil {
		t.Fatal(err)
	}
	if len(response.Models) != 1 || response.Models[0].ID != "claude-sonnet-4-6" {
		t.Fatalf("unexpected Anthropic models: %#v", response.Models)
	}
	if response.Models[0].DisplayName == nil || *response.Models[0].DisplayName != "Claude Sonnet 4.6" {
		t.Fatalf("display name not parsed: %#v", response.Models[0])
	}
}

func TestAIModelListCompatibleWithoutAPIKey(t *testing.T) {
	server := withAIModelListTestServer(t, func(w http.ResponseWriter, r *http.Request) {
		if r.URL.Path != "/v1/models" {
			t.Fatalf("unexpected path: %s", r.URL.Path)
		}
		if r.Header.Get("Authorization") != "" {
			t.Fatalf("compatible request should not send empty auth header")
		}
		_, _ = w.Write([]byte(`{"data":[{"id":"custom-model"}]}`))
	})

	response, err := listAIModels(context.Background(), aiModelListRequest{
		ProviderType: aiProviderTypeOpenAICompatible,
		BaseURL:      server.URL + "/v1/",
		APIKey:       "",
	}, localeZhCN)
	if err != nil {
		t.Fatal(err)
	}
	if len(response.Models) != 1 || response.Models[0].ID != "custom-model" {
		t.Fatalf("unexpected compatible models: %#v", response.Models)
	}
}

func TestAIModelListProviderErrorReturnsRawProviderResponse(t *testing.T) {
	server := withAIModelListTestServer(t, func(w http.ResponseWriter, r *http.Request) {
		w.Header().Set("Content-Type", "text/plain")
		w.WriteHeader(http.StatusUnauthorized)
		_, _ = w.Write([]byte(`invalid sk-test-secret`))
	})

	_, err := listAIModels(context.Background(), aiModelListRequest{
		ProviderType: aiProviderTypeOpenAI,
		BaseURL:      server.URL + "/v1",
		APIKey:       "sk-test-secret",
	}, localeZhCN)
	var httpErr *aiModelListHTTPError
	if !errors.As(err, &httpErr) {
		t.Fatalf("expected aiModelListHTTPError, got %#v", err)
	}
	if httpErr.status != http.StatusUnauthorized || httpErr.reason != "http_401" {
		t.Fatalf("provider 401 should pass through with reason http_401, got status=%d reason=%q", httpErr.status, httpErr.reason)
	}
	if httpErr.message == nil || *httpErr.message != "invalid sk-test-secret" {
		t.Fatalf("provider message should keep raw body: %#v", httpErr.message)
	}
	if httpErr.providerResponse == nil || httpErr.providerResponse.Body == nil || *httpErr.providerResponse.Body != "invalid sk-test-secret" {
		t.Fatalf("provider response body should keep raw body: %#v", httpErr.providerResponse)
	}
	if httpErr.providerResponse.Status == nil || *httpErr.providerResponse.Status != http.StatusUnauthorized {
		t.Fatalf("provider status not captured: %#v", httpErr.providerResponse)
	}
	if httpErr.providerResponse.StatusText == nil || *httpErr.providerResponse.StatusText != "Unauthorized" {
		t.Fatalf("provider status text not captured: %#v", httpErr.providerResponse)
	}
	if got := httpErr.providerResponse.Headers["Content-Type"]; !strings.Contains(got, "text/plain") {
		t.Fatalf("provider response headers not captured: %#v", httpErr.providerResponse.Headers)
	}
}

func TestAIModelListProviderStatusPassthrough(t *testing.T) {
	cases := []struct {
		name           string
		providerStatus int
		wantReason     string
		wantStatus     int
	}{
		{name: "forbidden provider credentials", providerStatus: http.StatusForbidden, wantReason: "http_403", wantStatus: http.StatusForbidden},
		{name: "missing provider endpoint", providerStatus: http.StatusNotFound, wantReason: "http_404", wantStatus: http.StatusNotFound},
		{name: "provider validation error", providerStatus: http.StatusUnprocessableEntity, wantReason: "http_422", wantStatus: http.StatusUnprocessableEntity},
		{name: "provider rate limit", providerStatus: http.StatusTooManyRequests, wantReason: "http_429", wantStatus: http.StatusTooManyRequests},
		{name: "provider server error", providerStatus: http.StatusServiceUnavailable, wantReason: "http_503", wantStatus: http.StatusServiceUnavailable},
	}

	for _, tc := range cases {
		t.Run(tc.name, func(t *testing.T) {
			server := withAIModelListTestServer(t, func(w http.ResponseWriter, r *http.Request) {
				w.WriteHeader(tc.providerStatus)
				_, _ = w.Write([]byte(`provider failed`))
			})

			_, err := listAIModels(context.Background(), aiModelListRequest{
				ProviderType: aiProviderTypeOpenAI,
				BaseURL:      server.URL + "/v1",
				APIKey:       "sk-test-secret",
			}, localeZhCN)
			var httpErr *aiModelListHTTPError
			if !errors.As(err, &httpErr) {
				t.Fatalf("expected aiModelListHTTPError, got %#v", err)
			}
			if httpErr.status != tc.wantStatus || httpErr.reason != tc.wantReason {
				t.Fatalf("unexpected mapped error: status=%d reason=%q", httpErr.status, httpErr.reason)
			}
			if httpErr.providerResponse == nil || httpErr.providerResponse.Body == nil || *httpErr.providerResponse.Body != "provider failed" {
				t.Fatalf("provider response body not captured: %#v", httpErr.providerResponse)
			}
		})
	}
}

func TestAIModelListInvalidJSONReturnsBadRequest(t *testing.T) {
	server := withAIModelListTestServer(t, func(w http.ResponseWriter, r *http.Request) {
		_, _ = w.Write([]byte(`not-json`))
	})

	_, err := listAIModels(context.Background(), aiModelListRequest{
		ProviderType: aiProviderTypeOpenAI,
		BaseURL:      server.URL + "/v1",
		APIKey:       "sk-test-secret",
	}, localeZhCN)
	var httpErr *aiModelListHTTPError
	if !errors.As(err, &httpErr) || httpErr.status != http.StatusBadRequest || httpErr.code != "AI_MODEL_LIST_INVALID_JSON" {
		t.Fatalf("expected invalid JSON to map to 400, got %#v", err)
	}
	if httpErr.providerResponse == nil || httpErr.providerResponse.Body == nil || *httpErr.providerResponse.Body != "not-json" {
		t.Fatalf("invalid JSON should include raw provider body: %#v", httpErr.providerResponse)
	}
}

func TestAIModelListLargeProviderResponseReturnsPayloadTooLarge(t *testing.T) {
	server := withAIModelListTestServer(t, func(w http.ResponseWriter, r *http.Request) {
		_, _ = w.Write([]byte(strings.Repeat("x", aiModelListResponseBytes+1)))
	})

	_, err := listAIModels(context.Background(), aiModelListRequest{
		ProviderType: aiProviderTypeOpenAI,
		BaseURL:      server.URL + "/v1",
		APIKey:       "sk-test-secret",
	}, localeZhCN)
	var httpErr *aiModelListHTTPError
	if !errors.As(err, &httpErr) || httpErr.status != http.StatusRequestEntityTooLarge || httpErr.code != "AI_MODEL_LIST_RESPONSE_TOO_LARGE" {
		t.Fatalf("expected large provider response to return 413, got %#v", err)
	}
	if httpErr.providerResponse != nil {
		t.Fatalf("oversized response must not return partial provider body: %#v", httpErr.providerResponse)
	}
}

func TestAIModelListTimeout(t *testing.T) {
	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		time.Sleep(50 * time.Millisecond)
	}))
	previousClient := aiModelListHTTPClient
	aiModelListHTTPClient = &http.Client{Timeout: time.Millisecond}
	t.Cleanup(func() {
		aiModelListHTTPClient = previousClient
		server.Close()
	})

	_, err := listAIModels(context.Background(), aiModelListRequest{
		ProviderType: aiProviderTypeOpenAI,
		BaseURL:      server.URL + "/v1",
		APIKey:       "sk-test-secret",
	}, localeZhCN)
	var httpErr *aiModelListHTTPError
	if !errors.As(err, &httpErr) || httpErr.status != http.StatusRequestTimeout || httpErr.code != "AI_MODEL_LIST_TIMEOUT" {
		t.Fatalf("expected timeout error, got %#v", err)
	}
}

File: packages/cloudflare/src/ai-models.test.ts
// Worker AI 模型列表测试保护认证代理、provider 形状归一和原始错误回显，避免请求 API key 进入响应 headers。
import { beforeEach, describe, expect, it, vi } from "vitest";
import { listAIModels } from "./ai-models";
import type { Env } from "./types";

const authMocks = vi.hoisted(() => ({
  requireAuth: vi.fn(),
}));

vi.mock("./auth", () => ({
  requireAuth: authMocks.requireAuth,
}));

const authUser = {
  id: "usr_models",
  email: "models@example.com",
  name: "Model User",
  role: "admin" as const,
  banned: 0,
  ban_reason: "",
  password_hash: "hash",
  reset_token_hash: null,
  reset_token_expires_at: null,
  created_at: "2026-06-05T00:00:00.000Z",
  updated_at: "2026-06-05T00:00:00.000Z",
};

function envFixture(): Env {
  return { DB: {} as D1Database, ASSETS_BUCKET: {} as R2Bucket };
}

function requestFor(body: unknown): Request {
  return new Request("https://renewlet.test/api/app/ai/models/list", {
    method: "POST",
    headers: {
      authorization: "Bearer test",
      "content-type": "application/json",
      "x-renewlet-locale": "zh-CN",
    },
    body: JSON.stringify(body),
  });
}

describe("Cloudflare AI model list proxy", () => {
  beforeEach(() => {
    authMocks.requireAuth.mockReset();
    authMocks.requireAuth.mockResolvedValue({ user: authUser, session: { id: "ses" }, token: "test" });
    vi.unstubAllGlobals();
  });

  it("loads and normalizes OpenAI shape models", async () => {
    const fetchMock = vi.fn().mockResolvedValue(new Response(JSON.stringify({
      data: [
        { id: "gpt-5.1", created: 1780000000, owned_by: "openai" },
        { id: "gpt-5.1", created: 1780000000, owned_by: "openai" },
      ],
    }), { status: 200 }));
    vi.stubGlobal("fetch", fetchMock);

    const response = await listAIModels(requestFor({
      providerType: "openai",
      baseUrl: "",
      apiKey: "sk-test-secret",
    }), envFixture());
    const body = await response.json() as { models: Array<{ id: string; ownedBy: string | null }>; truncated: boolean };

    expect(fetchMock).toHaveBeenCalledWith("https://api.openai.com/v1/models", expect.objectContaining({
      method: "GET",
      headers: expect.any(Headers),
    }));
    expect((fetchMock.mock.calls[0]?.[1] as { headers: Headers }).headers.get("authorization")).toBe("Bearer sk-test-secret");
    expect(body.models).toEqual([expect.objectContaining({ id: "gpt-5.1", ownedBy: "openai" })]);
    expect(body.truncated).toBe(false);
  });

  it("uses Gemini model ids and filters non generateContent models", async () => {
    vi.stubGlobal("fetch", vi.fn().mockResolvedValue(new Response(JSON.stringify({
      models: [
        {
          name: "models/gemini-2.5-pro",
          baseModelId: "gemini-2.5-pro",
          displayName: "Gemini 2.5 Pro",
          inputTokenLimit: 1048576,
          outputTokenLimit: 65536,
          supportedGenerationMethods: ["generateContent"],
          thinking: true,
        },
        {
          name: "models/text-embedding",
          supportedGenerationMethods: ["embedContent"],
        },
      ],
    }), { status: 200 })));

    const response = await listAIModels(requestFor({
      providerType: "gemini",
      baseUrl: "",
      apiKey: "AIza-test-secret",
    }), envFixture());
    const body = await response.json() as { models: Array<{ id: string; displayName: string | null; inputTokenLimit: number | null; capabilities: { thinking: boolean | null } }> };

    expect(body.models).toEqual([expect.objectContaining({
      id: "gemini-2.5-pro",
      displayName: "Gemini 2.5 Pro",
      inputTokenLimit: 1048576,
      capabilities: expect.objectContaining({ thinking: true }),
    })]);
  });

  it("loads Claude models with Anthropic headers", async () => {
    const fetchMock = vi.fn().mockResolvedValue(new Response(JSON.stringify({
      data: [{ id: "claude-sonnet-4-6", display_name: "Claude Sonnet 4.6", created_at: "2026-01-01T00:00:00Z", type: "model" }],
    }), { status: 200 }));
    vi.stubGlobal("fetch", fetchMock);

    const response = await listAIModels(requestFor({
      providerType: "anthropic",
      baseUrl: "",
      apiKey: "sk-ant-test-secret",
    }), envFixture());
    const body = await response.json() as { models: Array<{ id: string; displayName: string | null }> };
    const headers = (fetchMock.mock.calls[0]?.[1] as { headers: Headers }).headers;

    expect(fetchMock.mock.calls[0]?.[0]).toBe("https://api.anthropic.com/v1/models");
    expect(headers.get("x-api-key")).toBe("sk-ant-test-secret");
    expect(headers.get("anthropic-version")).toBe("2023-06-01");
    expect(body.models[0]).toMatchObject({ id: "claude-sonnet-4-6", displayName: "Claude Sonnet 4.6" });
  });

  it("supports OpenAI compatible lists without requiring an API key", async () => {
    const fetchMock = vi.fn().mockResolvedValue(new Response(JSON.stringify({
      data: [{ id: "custom-model" }],
    }), { status: 200 }));
    vi.stubGlobal("fetch", fetchMock);

    const response = await listAIModels(requestFor({
      providerType: "openai-compatible",
      baseUrl: "https://llm.example.com/v1/",
      apiKey: "",
    }), envFixture());
    const body = await response.json() as { models: Array<{ id: string }> };

    expect(fetchMock.mock.calls[0]?.[0]).toBe("https://llm.example.com/v1/models");
    expect((fetchMock.mock.calls[0]?.[1] as { headers: Headers }).headers.get("authorization")).toBeNull();
    expect(body.models[0]?.id).toBe("custom-model");
  });

  it("returns raw provider response body for model list errors", async () => {
    vi.stubGlobal("fetch", vi.fn().mockResolvedValue(new Response("invalid sk-test-secret", {
      status: 401,
      headers: { "content-type": "text/plain" },
    })));

    await expect(listAIModels(requestFor({
      providerType: "openai",
      baseUrl: "",
      apiKey: "sk-test-secret",
    }), envFixture())).rejects.toMatchObject({
      status: 401,
      code: "AI_MODEL_LIST_FAILED",
      details: {
        reason: "http_401",
        providerMessage: "invalid sk-test-secret",
        providerResponse: {
          status: 401,
          headers: expect.objectContaining({ "content-type": "text/plain" }),
          body: "invalid sk-test-secret",
          bodyTruncated: false,
        },
      },
    });
  });

  it("passes through provider rate limits for display", async () => {
    vi.stubGlobal("fetch", vi.fn().mockResolvedValue(new Response("slow down", { status: 429 })));

    await expect(listAIModels(requestFor({
      providerType: "openai",
      baseUrl: "",
      apiKey: "sk-test-secret",
    }), envFixture())).rejects.toMatchObject({
      status: 429,
      code: "AI_MODEL_LIST_FAILED",
      details: {
        reason: "http_429",
        providerResponse: expect.objectContaining({ body: "slow down", status: 429 }),
      },
    });
  });

  it("passes through provider server errors for display", async () => {
    vi.stubGlobal("fetch", vi.fn().mockResolvedValue(new Response("provider down", { status: 503 })));

    await expect(listAIModels(requestFor({
      providerType: "openai",
      baseUrl: "",
      apiKey: "sk-test-secret",
    }), envFixture())).rejects.toMatchObject({
      status: 503,
      code: "AI_MODEL_LIST_FAILED",
      details: {
        reason: "http_503",
        providerResponse: expect.objectContaining({ body: "provider down", status: 503 }),
      },
    });
  });

  it("returns bad request for invalid provider JSON", async () => {
    vi.stubGlobal("fetch", vi.fn().mockResolvedValue(new Response("not-json", { status: 200 })));

    await expect(listAIModels(requestFor({
      providerType: "openai",
      baseUrl: "",
      apiKey: "sk-test-secret",
    }), envFixture())).rejects.toMatchObject({
      status: 400,
      code: "AI_MODEL_LIST_INVALID_JSON",
      details: {
        reason: "invalid_json",
        providerResponse: expect.objectContaining({ body: "not-json", status: 200 }),
      },
    });
  });

  it("returns payload too large for oversized provider responses", async () => {
    vi.stubGlobal("fetch", vi.fn().mockResolvedValue(new Response("x".repeat((1 << 20) + 1), { status: 200 })));

    await expect(listAIModels(requestFor({
      providerType: "openai",
      baseUrl: "",
      apiKey: "sk-test-secret",
    }), envFixture())).rejects.toMatchObject({
      status: 413,
      code: "AI_MODEL_LIST_RESPONSE_TOO_LARGE",
      details: { reason: "response_too_large" },
    });
  });

  it("returns a readable timeout error", async () => {
    vi.stubGlobal("fetch", vi.fn().mockRejectedValue(new DOMException("aborted", "AbortError")));

    await expect(listAIModels(requestFor({
      providerType: "gemini",
      baseUrl: "",
      apiKey: "AIza-test-secret",
    }), envFixture())).rejects.toMatchObject({
      status: 408,
      code: "AI_MODEL_LIST_TIMEOUT",
      details: { reason: "timeout" },
    });
  });
});
